[WIP] Expression evaluator and variable watches

rxlecky opened the pull request:
When debugger breaks execution, expressions can be evaluated in the current stack frame.

Closes #7057.

SeleckyErik commented:
![20190224_034627](https://user-images.githubusercontent.com/35656626/53294634-98fb6d00-37ea-11e9-8d63-f59a4f82d740.gif)

SeleckyErik commented:
Not quite sure what is the reason of the failing build...

akien-mga commented:
> Not quite sure what is the reason of the failing build...

A style issue as shown by this diff: https://travis-ci.org/godotengine/godot/jobs/497835792
Our `clang-format` configuration disallows having more than one consecutive newline. See [Style guide](http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html).

SeleckyErik commented:
Oh, I see! I knew it was coming from `clang-format`, just wasn't sure what the problem was. Thanks!

SeleckyErik commented:
Wait, this is strange. This is the reported error from `clang-format`:

```
*** The following differences were found between the code to commit and the clang-format rules:
--- a/core/script_debugger_remote.cpp	2019-02-24 18:09:18.317826952 +0000
+++ b/core/script_debugger_remote.cpp	2019-02-24 18:09:27.113682041 +0000
@@ -1067,7 +1067,6 @@
 	physics_frame_time = p_physics_frame_time;
 }
 
-
 ScriptDebuggerRemote::ResourceUsageFunc ScriptDebuggerRemote::resource_usage_func = NULL;
 
 ScriptDebuggerRemote::ScriptDebuggerRemote() :
*** Aborting, please fix your commit(s) with 'git commit --amend' or 'git rebase -i <hash>'
```

However, it doesn't seem to match the source. Not to mention that there is not double newline:
https://github.com/godotengine/godot/blob/5452a4d4ec183d1f99b3e2e30fe838e5010a25c1/core/script_debugger_remote.cpp#L1066-L1080

It seems as if `clang-format` couldn't see the whole `set_allow_focus_steal_pid` function definition.

Also, I did not touch this section of the code at all, so it's weird this error popped up now, and not when the code was created.

akien-mga commented:
> It seems as if `clang-format` couldn't see the whole `set_allow_focus_steal_pid` function definition.

Indeed, it's because it has been removed in the master branch in 9d78274e0, which is not in your own branch. Travis then applies your PR on top of the master branch, and creates this style issue when merging the changes together.

SeleckyErik commented:
Of course, that makes sense!

Thanks for the explanation, will remember it not to ask the same question again.

akien-mga commented:
> Thanks for the explanation, will remember it not to ask the same question again.

Well I don't mind your asking, that was a tricky case :D It took me a while to understand why it was showing this style error myself.

SeleckyErik commented:
True, it wasn't a trivial case. :) Well, good thing it builds now without errors.

rxlecky committed:
Added Expression Evaluator and Script Watches the debugger

Evaluator allows  expressions to be evaluated in the current stack frame. Watches keep track of multiple
expressions.

SeleckyErik commented:
Updated the previous version: Added script watches allowing tracking results of multiple expressions. A watch can also be locked or unlocked. When locked, the variables in the expression are locked to reference the variables in that they reference at the moment of locking. If not locked, the watch is always evaluated in the context of the current stack frame.

One more feature that is planned, but currently disabled is tracking watches - making watch tracking makes it break execution when the result of the watch changes.

SeleckyErik commented:
@bojidar-bg @vnen I was talking to Juan a while back and he told me either of you guys would hopefully find time to skim over my code and give me some feedback.

SeleckyErik commented:
Hey @vnen I was looking into GDScript `debug_get_stack_level_locals` function that gives me some trouble. It returns all local variables in the function that the given stack frame is in - including functions that were not yet declared or initialized. For those, it returns `Variant::NULL` as their value. I believe the correct behaviour would be not to return the variables that do not exist yet.

Looking into the function that this function relies on, I found this code:
https://github.com/godotengine/godot/blob/29a1202d39b0efb617a8ff4ee9764612f3e0124f/modules/gdscript/gdscript_function.cpp#L1669-L1676

Was the line `if (sd.line > p_line) break;` supposed to skip the variables that are not yet declared and is just failing somewhere or am I misunderstanding the code? Thanks for the help!

reduz commented on core/script_language.cpp:
```diff
@@ -360,6 +396,190 @@ void ScriptDebugger::clear_breakpoints() {
+	return watches[p_index].expression->get_error_text();
+}
+
+bool ScriptDebugger::evaluate_watches(int p_stack_level, int p_watch) {
```
Some feedback:
1. This function will be called for every line of script code being run, and it will make execution considerably slower so make sure nothing is run if no watches exist by using some sort of inline static bool to check watches.
2. This function can be called from multiple threads, make sure to add a mutex to it if it actually runs
3. You will need to check if waches use stack variables from the current function, if so they will need to be erased once function exits.

reduz commented on core/math/expression.h:
```diff
@@ -36,6 +36,12 @@
 class Expression : public Reference {
 	GDCLASS(Expression, Reference)
 public:
+	struct Context {
```
This is added here and then unused, it should probably be added somewhere else.
Also, dont store a pointer, you should instead store an ObjectID via object->get_instance_id() and then resolve to pointer using ObjectDB::get_instance(). This way if the object was erased, it will fail and you can remove the watch.

akien-mga commented:
@SeleckyErik Do you know how to edit the PR to take @reduz's feedback into account, or do you need more details?

SeleckyErik commented:
@akien-mga Oh, sorry, I forgot to write it down here. I was in touch with Juan already and I wrote down a checklist of what needs to be done with some details. However, I am currently on a short vacation. I will start working on finishing the PR (and the other open PRs too) sometime towards the end of the next week. 🙂

SeleckyErik commented:
I believe these commits address all pieces of feedback that I got from @reduz, including what we discussed on IRC.

SeleckyErik commented on core/script_debugger_remote.cpp:
```diff
@@ -505,7 +505,10 @@ void ScriptDebuggerRemote::line_poll() {
 
 	// Checking for stale watches
```
This whole thing should probably not happen on each line, but rather use the if statement above or some predefined frequency.

Reneator commented:
Really looking forward to this! *hypetrain*

akien-mga commented:
Needs a rebase to solve a merge conflict, otherwise it should be good to review in a PR meeting.

Calinou commented:
Similar to other places where source code is written, it would make sense to use the monospace/fixed-width font for the Expression output and LineEdit :slightly_smiling_face:

SeleckyErik commented:
@akien-mga When's the PR meeting taking place? There are a still some changes I want to make before this is merged.

SeleckyErik commented:
@Calinou Thanks, I will change that. :)

akien-mga commented:
@SeleckyErik What's the status on this PR? Are there still changes you want to make before it's merged?

SeleckyErik commented:
@akien-mga I'm sorry, I was only in touch with Juan and didn't update you. Currently I was working on finishing #27742. I will push the final iteration on it this week. This PR will have to wait, though, because next week I have final school week and I have to focus on finishing up all formalities. I will resume working on this PR in July. If that means that it won't make it into 3.2, I am sorry for the inconvenience, but I am pressed by deadlines right now.

akien-mga commented:
No worry, take your time :)

There's no planned date for 3.2's feature freeze for now, but it will likely towards the end of July, so you should have time to finish this one.

SeleckyErik commented:
Nice, that should be enough time indeed. I will keep you updated in case of any changes.

rxlecky commented:
Worked on the Evaluator part and addressed mainly UI/UX-related feedback:
- Updated font to monospace (thanks for a reminder @Calinou 🙂)
- Enabled scroll following on the expression log RichTextLabel
- Added colors to distinguish normal results from errors
- Implemented expression history, allowing access to previously evaluated expressions
- Added a way to clear expression log (UI is WIP)

![2019-08-15_00-35-04](https://user-images.githubusercontent.com/35656626/63061876-129d2980-bef7-11e9-94f6-939da9cc10d2.gif)
_GIF showing UI/UXchanges_

Also, evaluator can now be used regardless of whether debugger is running or not or whether a breakpoint was hit or not; It can evaluate GDScript public functions and constants without a debugger context. So besides debugging, it can also double for a pocket calculator. 🙂

Calinou commented:
@rxlecky According to the GIF, the expression log (and its input field) still seems to use the default font rather than the code font.

rxlecky commented:
@Calinou No, the GIF shows the current. Honestly, that's why I tagged you, because I find it hard to spot the differences between the fonts so I was hoping you'd give me a hand. Is this the correct way to set the code font?

https://github.com/godotengine/godot/blob/58af21e12a24d575510fa0a95b5d8d3c0d93c6f1/editor/expression_evaluator.cpp#L129-L138

Also, I did a small comparison, I put two LineEdits next to each other. The left one is the current LineEdit from this PR and the right one is a default LineEdit. There seems to be a small difference but still I am not sure whether the left one uses the correct code font.

![obrázok](https://user-images.githubusercontent.com/35656626/63104567-84b95100-bf7f-11e9-9e1a-acced860f345.png)

Calinou commented:
@rxlecky I got the font setting work as expected :slightly_smiling_face:

```diff
diff --git a/editor/expression_evaluator.cpp b/editor/expression_evaluator.cpp
index 16df7d7b4..df63360cc 100644
--- a/editor/expression_evaluator.cpp
+++ b/editor/expression_evaluator.cpp
@@ -33,6 +33,7 @@
 #include "core/math/expression.h"
 #include "core/os/input.h"
 #include "core/os/keyboard.h"
+#include "editor_node.h"
 #include "editor/editor_settings.h"
 #include "scene/gui/button.h"
 #include "scene/gui/control.h"
@@ -126,7 +127,7 @@ HistoryLineEdit::HistoryLineEdit() {
        history.index = 0;
        set_process_input(true);
 
-       Ref<Font> mono_font = get_font("source", "EditorFonts");
+       Ref<Font> mono_font = EditorNode::get_singleton()->get_gui_base()->get_font("source", "EditorFonts");
 
        set_h_size_flags(SIZE_EXPAND_FILL);
 
@@ -238,7 +239,7 @@ void ExpressionEvaluator::set_result(const String &p_result, bool p_error) {
 ExpressionEvaluator::ExpressionEvaluator() {
        expression = memnew(Expression);
 
-       Ref<Font> mono_font = get_font("source", "EditorFonts");
+       Ref<Font> mono_font = EditorNode::get_singleton()->get_gui_base()->get_font("source", "EditorFonts");
 
        set_name("Expressions");
        set_focus_mode(FOCUS_NONE);
```

I don't know why it solves the issue, but I noticed the same thing is done in https://github.com/godotengine/godot/blob/de8ce3e625e74833aec6a5d165e7e82100a1dbf3/editor/find_in_files.cpp#L522.

rxlecky commented:
Ahhh, I see, thanks for the help! I'll fix it. :)

akien-mga commented:
Waiting for #27742 to be merged as it will conflict with this PR, which should then be rebased (debugger moving from core to scene).

akien-mga commented:
Moving to 4.0 milestone as it still needs some work so it won't be ready to merge for the imminent 3.2 release.

aaronfranke commented:
@rxlecky Is this still desired? If so, it needs to be rebased on the latest master branch. #27742 previously blocked this PR, but that has since been merged.

If not, abandoned pull requests will be closed in the future as announced [here](https://godotengine.org/article/headsup-vulkan-merged-master-branch-unstable).

aaronfranke commented:
This PR has not received any new commits for almost a year and is still a WIP, closing. If this PR is still desired, it can be re-opened or re-created later if it is rebased and finished.

Ark2000 commented:
Do we have expression evaulator in debugger now?

Calinou commented:
> Do we have expression evaulator in debugger now?

No, not yet. It's being worked on in https://github.com/godotengine/godot/pull/60134.

akien-mga commented:
Superseded by #97647.

Calinou commented:
Note that https://github.com/godotengine/godot/pull/97647 doesn't implement variable watches. This would need to be reimplemented from scratch, given a lot of GDScript code has changed since this PR was last updated (along with the expression evaluator UI, which is closely related).
